# Confidence intervals in python

낮잠을 자는 유아와 낮잠을 자지 않는 유아 간의 다양한 수면 변수 차이를 조사한 연구 결과 데이터를 분석해보자.

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import t

In [53]:
df = pd.read_csv("nap_no_nap.csv") 
pd.set_option('display.max_columns', 30)
df

id     sex  age (months)  dlmo time  days napped  napping  \
0    1  female          33.7      19.24            0        0   
1    2  female          31.5      18.27            0        0   
2    3    male          31.9      19.14            0        0   
3    4  female          31.6      19.69            0        0   
4    5  female          33.0      19.52            0        0   
5    6  female          36.2      18.22            4        1   
6    7    male          36.3      19.28            1        1   
7    8    male          30.0      21.06            5        1   
8    9    male          33.2      19.38            2        1   
9   10  female          37.1      19.93            3        1   
10  11    male          32.9      18.79            4        1   
11  12  female          35.0      19.65            5        1   
12  13    male          35.1      19.83            3        1   
13  14  female          35.6      19.88            4        1   
14  15  female          36.6      19.94            4        1   
15  16    male          36.5      20.25            3        1   
16  17  female          33.7      20.33            5        1   
17  18    male          36.4      20.16            5        1   
18  19  female          33.6      19.68            3        1   
19  20    male          33.8      20.51            3        1   

    nap lights outl time  nap sleep onset  nap midsleep  nap sleep offset  \
0                    NaN              NaN           NaN               NaN   
1                    NaN              NaN           NaN               NaN   
2                    NaN              NaN           NaN               NaN   
3                    NaN              NaN           NaN               NaN   
4                    NaN              NaN           NaN               NaN   
5                  14.00            14.22         15.00             15.78   
6                  14.75            15.03         15.92             16.80   
7                  13.09            13.43         14.44             15.46   
8                  14.41            14.42         15.71             17.01   
9                  13.12            13.42         14.31             15.19   
10                 13.99            14.03         14.85             15.68   
11                 13.18            13.45         14.33             15.21   
12                 13.94            14.48         15.26             16.03   
13                 12.68            13.08         13.92             14.76   
14                 12.71            12.88         13.80             14.72   
15                 13.74            14.68         15.66             16.64   
16                 13.15            13.87         14.49             15.11   
17                 12.47            12.56         13.30             14.05   
18                 14.71            14.85         15.46             16.07   
19                 12.68            13.54         14.30             15.07   

    nap wake time  nap duration  nap time in bed  night bedtime  \
0             NaN           NaN              NaN          20.45   
1             NaN           NaN              NaN          19.23   
2             NaN           NaN              NaN          19.60   
3             NaN           NaN              NaN          19.46   
4             NaN           NaN              NaN          19.21   
5           16.28         93.75           137.00          19.95   
6           16.08        106.00            80.00          20.60   
7           15.82        121.60           163.80          22.01   
8           16.60        155.50           131.25          20.24   
9           15.30        106.67           130.67          20.78   
10          16.10         98.75           126.60          19.45   
11          15.35        105.80           130.40          20.18   
12          15.78         93.33           110.20          20.22   
13          15.00        100.75           139.33          20.26   
14          14.

sleep variables: Bedtime (lights-off time in decimalized time), Night Sleep Onset Time (in decimalized time), Wake Time (sleep end time in decimalized time), Night Sleep Duration (interval between sleep onset and sleep end in minutes), and Total 24-Hour Sleep Duration (in minutes).

### Average bedtime confidence interval for napping and non napping toddlers
낮잠을 자는 유아와 낮잠을 자지 않는 유아 평균 bedtime의 95% 신뢰구간을 만들어보자.

낮잠 자는 유아와 낮잠 자지 않는 유아의 bedtime 분리

In [42]:
bedtime_nap = df[df['napping']==1]['night bedtime']

In [43]:
bedtime_no_nap = df[df['napping']==0]['night bedtime']

1) mean

In [44]:
nap_mean_bedtime = np.mean(bedtime_nap)
nap_mean_bedtime

20.304

In [45]:
no_nap_mean_bedtime = np.mean(bedtime_no_nap)
no_nap_mean_bedtime

19.590000000000003

2) standard deviation

In [46]:
nap_s_bedtime = np.std(bedtime_nap, ddof=1)
nap_s_bedtime

0.5910619981984009

In [47]:
no_nap_s_bedtime = np.std(bedtime_no_nap, ddof=1)
no_nap_s_bedtime

0.5075923561284187

3) standard error

In [48]:
nap_se_mean_bedtime = nap_s_bedtime/np.sqrt(df[df['napping']==1].id.count())
nap_se_mean_bedtime

0.15261155170656565

In [49]:
no_nap_se_mean_bedtime = no_nap_s_bedtime/np.sqrt(df[df['napping']==0].id.count())
no_nap_se_mean_bedtime

0.22700220263248522

To build a 95% confidence interval, what is the value of t\*?  You can find this value using the percent point function (PPF): 
```
from scipy.stats import t

t.ppf(probability, df)
```
This will return the quantile value such that to the left of this value, the tail probabiliy is equal to the input probabiliy (for the specified degrees of freedom).   
   
   
   
   
degree of freedom:
```
nap_sample : n-1 = 15-1 = 14 

no_nap_sample : n-1 = 5-1 = 4 
```

In [50]:
nap_t_star = t.ppf(.95, df=14)
no_nap_t_star = t.ppf(.95, df=4)

Confidence Interval = 'Best estimate' +- 'Margin of error'

In [51]:
# nap
nap_ci_ub = nap_mean_bedtime + ( nap_t_star * nap_se_mean_bedtime )
nap_ci_lb = nap_mean_bedtime - ( nap_t_star * nap_se_mean_bedtime )
(nap_ci_lb, nap_ci_ub)

(20.035203727142896, 20.5727962728571)

In [52]:
# no nap
no_nap_ci_ub = no_nap_mean_bedtime + ( no_nap_t_star * no_nap_se_mean_bedtime )
no_nap_ci_lb = no_nap_mean_bedtime - ( no_nap_t_star * no_nap_se_mean_bedtime )
(no_nap_ci_lb, no_nap_ci_ub)

(19.106066084961693, 20.073933915038314)

# 해석

95% 신뢰도로, 낮잠을 자는 유아의 bedtime은 20.04 - 20.57로 추정된다.   
95% 신뢰도로, 낮잠을 자지 않는 유아의 bedtime은 19.11 - 20.07로 추정된다.   
   
두 신뢰 구간이 겹치는 부분이 있다.   
두 그룹의 bedtime이 다르다고 주장할 수 없다.   
가설 검정이 추가적으로 필요하다.